src/map/unit.c

```diff
@@ -959,71 +964,60 @@ static enum unit_dir unit_getdir(const struct block_list *bl)
-// flag:
-//  &1  Do not send position update packets.
-static int unit_blown(struct block_list *bl, int dx, int dy, int count, int flag)
+/**
+ * Pushes a unit by given amount of cells into given direction.
+ * @remark Only map cell restrictions are respected.
+ * @param bl: unit to push
+ * @param dir: direction of the push motion
+ * @param count: amount of cells to push
+ * @param no_update: don't send position update packets when true
+ * @return amount of pushed cells.
+ */
+static int unit_blown(struct block_list *bl, enum unit_dir dir, int count, bool no_update)
 {
-	if(count) {
-		struct map_session_data* sd;
-		struct skill_unit* su = NULL;
-		int nx, ny, result;
-
-		nullpo_ret(bl);
-
-		sd = BL_CAST(BL_PC, bl);
-		su = BL_CAST(BL_SKILL, bl);
-
-		result = path->blownpos(bl, bl->m, bl->x, bl->y, dx, dy, count);
-
-		nx = result>>16;
-		ny = result&0xffff;
+	if (count == 0)
+		return 0;
+	nullpo_ret(bl);
 
-		if(!su) {
-			unit->stop_walking(bl, STOPWALKING_FLAG_NONE);
-		}
+	int result = path->blownpos(bl, bl->m, bl->x, bl->y, dir, count);
+	int pushto_x = result >> 16;
+	int pushto_y = result & 0xffff;
 
-		if( sd ) {
-			unit->stop_stepaction(bl); //Stop stepaction when knocked back
-			sd->ud.to_x = nx;
-			sd->ud.to_y = ny;
-		}
+	struct skill_unit* su = BL_CAST(BL_SKILL, bl);
+	if (su == NULL)
+		unit->stop_walking(bl, STOPWALKING_FLAG_NONE);
 
-		dx = nx-bl->x;
-		dy = ny-bl->y;
+	struct map_session_data* sd = BL_CAST(BL_PC, bl);
+	if (sd != NULL) {
+		unit->stop_stepaction(bl);
+		sd->ud.to_x = pushto_x;
+		sd->ud.to_y = pushto_y;
+	}
 
-		if(dx || dy) {
-			map->foreachinmovearea(clif->outsight, bl, AREA_SIZE, dx, dy, bl->type == BL_PC ? BL_ALL : BL_PC, bl);
+	int delta_x = pushto_x - bl->x;
+	int delta_y = pushto_y - bl->y;
 
-			if(su) {
-				skill->unit_move_unit_group(su->group, bl->m, dx, dy);
-			} else {
-				map->moveblock(bl, nx, ny, timer->gettick());
-			}
+	if (delta_x != 0 || delta_y != 0) {
+		enum bl_type receiver_type = bl->type == BL_PC ? BL_ALL : BL_PC;
+		map->foreachinmovearea(clif->outsight, bl, AREA_SIZE, delta_x, delta_y, receiver_type, bl);
 
-			map->foreachinmovearea(clif->insight, bl, AREA_SIZE, -dx, -dy, bl->type == BL_PC ? BL_ALL : BL_PC, bl);
+		if (su != NULL)
+			skill->unit_move_unit_group(su->group, bl->m, delta_x, delta_y);
+		else
+			map->moveblock(bl, pushto_x, pushto_y, timer->gettick());
 
-			if(!(flag&1)) {
-				clif->blown(bl);
-			}
+		map->foreachinmovearea(clif->insight, bl, AREA_SIZE, -delta_x, -delta_y, receiver_type, bl);
 
-			if(sd) {
-				if(sd->touching_id) {
-					npc->touchnext_areanpc(sd, false);
-				}
-				if (map->getcell(bl->m, bl, bl->x, bl->y, CELL_CHKNPC)) {
-					npc->touch_areanpc(sd, bl->m, bl->x, bl->y);
-				} else {
-					npc->untouch_areanpc(sd, bl->m, bl->x, bl->y);;
-				}
-			}
-		}
+		if (!no_update)
+			clif->blown(bl);
 
-		count = path->distance(dx, dy);
+		if (sd != NULL)
+			npc->handle_touch_events(sd, bl, bl->x, bl->y);
```

Comment:
The original code doesn't check for `bl->prev == NULL`. Don't know if simply calling `npc->handle_touch_events()` here will cause unexpected behaviour in case of `bl->prev` being `NULL`.

Reply:
Added flag for npc->handle_touch_events to opt in to that check.